File: src/data_pipelines/admin_lab_api/schemas/catalogue.py
"""Request/response models for routers/catalogue.py (admin-lab-plan.md §3.2)."""

from datetime import datetime

from pydantic import BaseModel

from data_pipelines.db.status import LessonStatus


class SpeakerWrite(BaseModel):
    name_he: str
    name_en: str
    slug: str


class SpeakerRead(BaseModel):
    """Built explicitly (not from_attributes) — lesson_count is a query-side aggregate,
    not a Speaker attribute."""

    id: int
    name_he: str
    name_en: str
    slug: str
    # Lessons, not series: a speaker no longer owns series (database-schema.md §3.2),
    # so "how much of this speaker is in the catalogue" is only answerable per lesson.
    lesson_count: int


class SpeakerBrief(BaseModel):
    """A speaker as it appears attached to something else — a series' derived speaker
    list, or a lesson's attribution."""

    id: int
    name_he: str
    name_en: str
    slug: str


class LessonTypeRead(BaseModel):
    id: int
    slug: str
    name_he: str
    name_en: str


class SeriesWrite(BaseModel):
    """No speaker and no source.

    A series' speakers are derived from its lessons, so there is nothing to write. Its
    source wiring is an `ingest_rules` row, which is created by accepting a surveyed
    playlist rather than by typing an adapter key into a form — see
    adding-series-plan.md §4. This form edits what a series *is*: its names, its
    slug, its subject."""

    name_he: str
    name_en: str
    slug: str
    lesson_type: str  # lesson_type slug
    description_he: str | None = None
    description_en: str | None = None


class SeriesRead(BaseModel):
    """Built explicitly, same reasoning as SpeakerRead — speakers and lesson_count are
    joined/aggregated, not Series attributes."""

    id: int
    name_he: str
    name_en: str
    slug: str
    lesson_type: str
    lesson_type_name_he: str
    description_he: str | None
    description_en: str | None
    lesson_count: int
    # From the series_speakers view, most lessons first. **Zero, one, or many** — empty
    # for a series discovered but not yet populated, several for an anthology. A UI
    # wanting "the" speaker takes the first; it must not assume there is one.
    speakers: list[SpeakerBrief]


class LessonRead(BaseModel):
    """Built explicitly (not from_attributes) — status is computed by
    db.status.lesson_status(), not a Lesson attribute."""

    id: int
    external_id: str
    url: str
    title_he: str
    title_en: str | None
    lesson_type: str | None
    speakers: list[SpeakerBrief]
    published_at: datetime | None
    recorded_at: datetime | None
    discovered_at: datetime
    status: LessonStatus


class ResetPreview(BaseModel):
    lesson_count: int
    warning: str


class ResetResult(BaseModel):
    deleted_count: int

File: frontend/src/api/catalogue.ts
import type { components } from './schema'
import { apiFetch } from './client'

export type Speaker = components['schemas']['SpeakerRead']
export type SpeakerWrite = components['schemas']['SpeakerWrite']
/** A speaker as it hangs off something else — a series' derived list, a lesson's
 *  attribution. Zero, one or several; never assume one. */
export type SpeakerBrief = components['schemas']['SpeakerBrief']
export type LessonType = components['schemas']['LessonTypeRead']
export type Series = components['schemas']['SeriesRead']
export type SeriesWrite = components['schemas']['SeriesWrite']
export type Lesson = components['schemas']['LessonRead']
export type LessonStatus = components['schemas']['LessonStatus']
export type ResetPreview = components['schemas']['ResetPreview']
export type ResetResult = components['schemas']['ResetResult']

export const listSpeakers = () => apiFetch<Speaker[]>('/api/speakers')
export const createSpeaker = (body: SpeakerWrite) =>
  apiFetch<Speaker>('/api/speakers', { method: 'POST', body: JSON.stringify(body) })
export const updateSpeaker = (id: number, body: SpeakerWrite) =>
  apiFetch<Speaker>(`/api/speakers/${id}`, { method: 'PUT', body: JSON.stringify(body) })
export const deleteSpeaker = (id: number) =>
  apiFetch<void>(`/api/speakers/${id}`, { method: 'DELETE' })

export const listLessonTypes = () => apiFetch<LessonType[]>('/api/lesson-types')

/** Filtering by speaker goes through the lessons a speaker actually teaches — a series
 *  has no speaker of its own (database-schema.md §3.2). */
export const listSeries = (speakerId?: number) =>
  apiFetch<Series[]>(speakerId ? `/api/series?speaker_id=${speakerId}` : '/api/series')
export const getSeries = (id: number) => apiFetch<Series>(`/api/series/${id}`)
export const createSeries = (body: SeriesWrite) =>
  apiFetch<Series>('/api/series', { method: 'POST', body: JSON.stringify(body) })
export const updateSeries = (id: number, body: SeriesWrite) =>
  apiFetch<Series>(`/api/series/${id}`, { method: 'PUT', body: JSON.stringify(body) })
export const deleteSeries = (id: number) => apiFetch<void>(`/api/series/${id}`, { method: 'DELETE' })

export const listSeriesLessons = (seriesId: number) =>
  apiFetch<Lesson[]>(`/api/series/${seriesId}/lessons`)
export const previewResetSeries = (seriesId: number) =>
  apiFetch<ResetPreview>(`/api/series/${seriesId}/reset`)
export const resetSeries = (seriesId: number) =>
  apiFetch<ResetResult>(`/api/series/${seriesId}/reset`, { method: 'POST' })
